File: internal/data/attention.go
// Package data — attention.go scans Copilot CLI session-state directories
// to determine real-time attention status for each session. This avoids
// relying on session-store.db which is only updated during full reindex.
package data

import (
	"bytes"
	"encoding/json"
	"io"
	"log/slog"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"time"

	"github.com/jongio/dispatch/internal/platform"
)

// sessionStateRel is the relative path from the user home directory to
// the Copilot CLI session-state directory.
const sessionStateRel = ".copilot/session-state"

// lastChunkSize is the number of bytes read from the end of events.jsonl
// to find the last complete event line. Events are typically 200–500 bytes,
// so 4 KB provides ample margin.
const lastChunkSize = 4096

// sessionEvent is a minimal representation of a Copilot CLI event from
// events.jsonl. Only the fields needed for attention classification are
// decoded.
type sessionEvent struct {
	Type      string `json:"type"`
	Timestamp string `json:"timestamp"`
}

// maxLockFileSize is the maximum size in bytes accepted for a lock file.
// Lock files contain a PID as an ASCII decimal string, which for the
// maximum PID of 4194304 is 7 bytes. 32 provides generous margin for
// whitespace/newlines while rejecting anything suspiciously large.
const maxLockFileSize = 32

// maxLockFilesPerSession caps the number of lock files checked per session
// directory to prevent resource exhaustion from adversarial lock file
// proliferation. Normal sessions have 0–1 lock files.
const maxLockFilesPerSession = 10

// deadSessionMaxAge is the maximum age of the last event for a dead session
// to be classified as AttentionWaiting. Older dead sessions are always Idle
// to avoid noise from long-abandoned sessions.
const deadSessionMaxAge = 24 * time.Hour

// interruptedMaxAge is the maximum age of the last event for a dead session
// with a stale lock to be classified as AttentionInterrupted. Older sessions
// are always Idle to avoid surfacing long-abandoned crashes.
const interruptedMaxAge = 72 * time.Hour

// ScanAttention reads the Copilot CLI session-state directories and returns
// a map of session ID → AttentionStatus. The threshold parameter controls
// how long a running session can be quiet before it is classified as stale.
//
// The scan is read-only and does not modify any files. It completes in
// under 50 ms for 100 sessions on typical hardware.
func ScanAttention(threshold time.Duration, workspaceRecovery bool) map[string]AttentionStatus {
	stateDir := sessionStatePath()
	if stateDir == "" {
		return nil
	}

	entries, err := os.ReadDir(stateDir)
	if err != nil {
		return nil
	}

	result := make(map[string]AttentionStatus, len(entries))

	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		sessionID := e.Name()
		dir := filepath.Join(stateDir, sessionID)

		status := classifySession(dir, threshold, workspaceRecovery)
		result[sessionID] = status
	}

	return result
}

// ScanAttentionQuick performs a fast first pass that only checks lock files
// for live sessions. Dead sessions are classified as AttentionIdle without
// reading events.jsonl. Use this for the initial scan to get dots visible
// immediately, then follow up with ScanAttention for full classification.
func ScanAttentionQuick(threshold time.Duration, workspaceRecovery bool) map[string]AttentionStatus {
	stateDir := sessionStatePath()
	if stateDir == "" {
		return nil
	}

	entries, err := os.ReadDir(stateDir)
	if err != nil {
		return nil
	}

	result := make(map[string]AttentionStatus, len(entries))

	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		sessionID := e.Name()
		dir := filepath.Join(stateDir, sessionID)

		pidRes := findSessionPID(dir)
		if pidRes.pid <= 0 {
			if pidRes.hasStale && workspaceRecovery {
				// Preliminary: mark interrupted so dots appear immediately.
				// Full scan refines this (e.g., turn_end → Waiting).
				result[sessionID] = AttentionInterrupted
			} else {
				// Dead session — skip events.jsonl for speed.
				result[sessionID] = AttentionIdle
			}
			continue
		}

		// Live session — full classification (fast since events.jsonl
		// is typically small for active sessions).
		result[sessionID] = classifyLiveSession(dir, threshold)
	}

	return result
}

// classifySession determines the attention status of a single session
// by checking its lock file and last event.
func classifySession(dir string, threshold time.Duration, workspaceRecovery bool) AttentionStatus {
	result := findSessionPID(dir)
	if result.pid > 0 {
		return classifyLiveSession(dir, threshold)
	}

	// No live process — check last event.
	eventsPath := filepath.Join(dir, "events.jsonl")
	evt, err := readLastEvent(eventsPath)
	if err != nil {
		return AttentionIdle
	}

	eventTime := parseEventTime(evt.Timestamp)
	if eventTime.IsZero() {
		return AttentionIdle
	}

	// Stale lock (dead PID) with workspace recovery enabled.
	if result.hasStale && workspaceRecovery {
		if time.Since(eventTime) > interruptedMaxAge {
			return AttentionIdle
		}
		switch {
		case strings.HasPrefix(evt.Type, "assistant.turn_end"),
			strings.HasPrefix(evt.Type, "assistant.message"):
			return AttentionWaiting
		default:
			return AttentionInterrupted
		}
	}

	// No stale lock (or feature disabled) — original dead-session logic.
	if time.Since(eventTime) > deadSessionMaxAge {
		return AttentionIdle
	}
	switch {
	case strings.HasPrefix(evt.Type, "assistant.turn_end"),
		strings.HasPrefix(evt.Type, "assistant.message"):
		return AttentionWaiting
	default:
		return AttentionIdle
	}
}

// classifyLiveSession classifies a session that has a live process.
func classifyLiveSession(dir string, threshold time.Duration) AttentionStatus {
	// Session has a live process — read the last event to classify.
	eventsPath := filepath.Join(dir, "events.jsonl")
	evt, err := readLastEvent(eventsPath)
	if err != nil {
		return AttentionStale // can't read events but session is running
	}

	// Parse event timestamp to check recency.
	eventTime := parseEventTime(evt.Timestamp)
	if !eventTime.IsZero() && time.Since(eventTime) > threshold {
		return AttentionStale
	}

	// Classify by event type.
	switch {
	case strings.HasPrefix(evt.Type, "assistant.turn_end"),
		strings.HasPrefix(evt.Type, "assistant.message"):
		return AttentionWaiting
	case strings.HasPrefix(evt.Type, "assistant.turn_start"),
		strings.HasPrefix(evt.Type, "tool.execution"),
		strings.HasPrefix(evt.Type, "hook."),
		strings.HasPrefix(evt.Type, "subagent."),
		strings.HasPrefix(evt.Type, "session.plan_changed"):
		return AttentionActive
	case evt.Type == "session.shutdown":
		return AttentionIdle
	default:
		// user.message, session.start, or unknown — AI hasn't started
		// responding yet, so treat as active.
		return AttentionActive
	}
}

// pidResult carries the outcome of a session lock-file check.
type pidResult struct {
	pid      int  // >0 if a live process owns the session
	hasStale bool // true if at least one lock file exists with a dead PID
}

// maxPID is a conservative upper bound for valid process IDs.
const maxPID = 4194304

// findSessionPID looks for inuse.*.lock files in the session directory.
// It returns a pidResult with the live PID if found, or hasStale=true
// if at least one lock file references a dead process.
func findSessionPID(dir string) pidResult {
	pattern := filepath.Join(dir, "inuse.*.lock")
	matches, err := filepath.Glob(pattern)
	if err != nil || len(matches) == 0 {
		return pidResult{}
	}

	hasStale := false
	for i, lockPath := range matches {
		if i >= maxLockFilesPerSession {
			slog.Warn("attention: too many lock files, skipping rest", "dir", dir, "count", len(matches))
			break
		}

		// Use Lstat to avoid following symlinks to FIFOs/devices that
		// could block ReadFile indefinitely (DoS vector).
		info, err := os.Lstat(lockPath)
		if err != nil || !info.Mode().IsRegular() {
			if err == nil {
				slog.Warn("attention: skipping non-regular lock file", "path", lockPath)
			}
			continue
		}
		if info.Size() >= maxLockFileSize {
			slog.Warn("attention: oversized lock file", "path", lockPath, "size", info.Size())
			continue
		}

		raw, err := os.ReadFile(lockPath)
		if err != nil {
			continue
		}

		pidStr := strings.TrimSpace(string(raw))

		// Validate: non-empty, pure ASCII digits only.
		if len(pidStr) == 0 {
			slog.Warn("attention: empty lock file", "path", lockPath)
			continue
		}
		valid := true
		for _, b := range pidStr {
			if b < '0' || b > '9' {
				valid = false
				break
			}
		}
		if !valid {
			slog.Warn("attention: non-numeric lock content", "path", lockPath, "content", pidStr)
			continue
		}

		pid, err := strconv.Atoi(pidStr)
		if err != nil || pid <= 0 || pid > maxPID {
			slog.Warn("attention: invalid PID in lock", "path", lockPath, "pid", pid)
			continue
		}

		if platform.IsProcessAlive(pid) {
			return pidResult{pid: pid}
		}
		hasStale = true
	}
	return pidResult{hasStale: hasStale}
}

// readLastEvent reads the last complete JSON line from an events.jsonl file
// using an O(1) seek-from-end strategy. It never reads the entire file.
func readLastEvent(path string) (sessionEvent, error) {
	var zero sessionEvent

	f, err := os.Open(path)
	if err != nil {
		return zero, err
	}
	defer f.Close()

	info, err := f.Stat()
	if err != nil {
		return zero, err
	}

	size := info.Size()
	if size == 0 {
		return zero, io.EOF
	}

	// Read the last chunk of the file.
	chunkSize := int64(lastChunkSize)
	if chunkSize > size {
		chunkSize = size
	}

	offset := size - chunkSize
	if _, err := f.Seek(offset, io.SeekStart); err != nil {
		return zero, err
	}

	buf := make([]byte, chunkSize)
	n, err := io.ReadFull(f, buf)
	if err != nil && err != io.ErrUnexpectedEOF {
		return zero, err
	}
	buf = buf[:n]

	// Find the last complete line (skip trailing empty line after final newline).
	buf = bytes.TrimRight(buf, "\n\r")
	lastNL := bytes.LastIndexByte(buf, '\n')
	var lastLine []byte
	if lastNL >= 0 {
		lastLine = buf[lastNL+1:]
	} else {
		// Only one line (or partial) in the chunk — use the entire buffer.
		lastLine = buf
	}

	var evt sessionEvent
	if err := json.Unmarshal(lastLine, &evt); err != nil {
		return zero, err
	}
	return evt, nil
}

// parseEventTime parses an ISO 8601 timestamp from a Copilot CLI event.
func parseEventTime(ts string) time.Time {
	if ts == "" {
		return time.Time{}
	}
	// Try RFC 3339 first (Go's standard ISO 8601 parser).
	t, err := time.Parse(time.RFC3339, ts)
	if err != nil {
		// Try with millisecond precision.
		t, err = time.Parse("2006-01-02T15:04:05.000Z", ts)
		if err != nil {
			return time.Time{}
		}
	}
	return t
}

// sessionStatePath returns the absolute path to the Copilot CLI
// session-state directory (~/.copilot/session-state/). Returns empty
// string if the home directory cannot be resolved.
func sessionStatePath() string {
	if override := os.Getenv("DISPATCH_SESSION_STATE"); override != "" {
		p := filepath.Clean(override)
		if runtime.GOOS == "windows" && strings.HasPrefix(p, `\\`) {
			return "" // reject UNC paths
		}
		return p
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	return filepath.Join(home, sessionStateRel)
}

File: electron/src/main/attention.ts
import { readdir, readFile, stat, open } from 'fs/promises';
import { homedir } from 'os';
import { join } from 'path';
import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type AttentionStatus =
  | 'idle'
  | 'stale'
  | 'active'
  | 'waiting'
  | 'interrupted'
  | 'working'
  | 'thinking'
  | 'compacting';

interface SessionEvent {
  type: string;
  timestamp: string;
}

interface PidResult {
  pid: number; // >0 if a live process owns the session
  hasStale: boolean; // true if at least one lock file references a dead PID
}

// ---------------------------------------------------------------------------
// Constants — mirror the Go implementation
// ---------------------------------------------------------------------------

const SESSION_STATE_REL = '.copilot/session-state';
const LAST_CHUNK_SIZE = 4096;
const MAX_LOCK_FILE_SIZE = 32;
const MAX_LOCK_FILES_PER_SESSION = 10;
const MAX_PID = 4_194_304;

// Event type prefixes
const EVENT_TURN_END = 'assistant.turn_end';
const EVENT_MESSAGE = 'assistant.message';
const EVENT_TURN_START = 'assistant.turn_start';
const EVENT_TOOL_EXECUTION = 'tool.execution';
const EVENT_HOOK = 'hook.';
const EVENT_SUBAGENT = 'subagent.';
const EVENT_PLAN_CHANGED = 'session.plan_changed';
const EVENT_SKILL_INVOKED = 'skill.invoked';
const EVENT_SHUTDOWN = 'session.shutdown';
const EVENT_ABORT = 'abort';
const EVENT_MODEL_CHANGE = 'session.model_change';
const EVENT_SYSTEM_MESSAGE = 'system.message';
const EVENT_COMPACTION = 'session.compaction_complete';

// Age thresholds
const DEAD_SESSION_MAX_AGE_MS = 24 * 60 * 60 * 1000; // 24 hours
const INTERRUPTED_MAX_AGE_MS = 72 * 60 * 60 * 1000; // 72 hours

// Session ID validation pattern — matches Go's validate.SessionIDPattern
const SESSION_ID_PATTERN = /^[a-zA-Z0-9][a-zA-Z0-9._-]{0,127}$/;

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Scans all session-state directories and returns a map of session ID to
 * attention status. Mirrors the Go ScanAttention function.
 *
 * @param thresholdMs - Stale threshold in milliseconds (default: 5 minutes)
 * @param workspaceRecovery - Enable interrupted detection for stale locks
 */
export async function scanAttention(
  thresholdMs = 5 * 60 * 1000,
  workspaceRecovery = true,
): Promise<Record<string, AttentionStatus>> {
  const stateDir = getSessionStatePath();
  if (!stateDir) return {};

  let entries: string[];
  try {
    const dirEntries = await readdir(stateDir, { withFileTypes: true });
    entries = dirEntries
      .filter((e) => e.isDirectory() && SESSION_ID_PATTERN.test(e.name))
      .map((e) => e.name);
  } catch {
    // Directory doesn't exist or isn't readable
    return {};
  }

  const result: Record<string, AttentionStatus> = {};

  // Process sessions in parallel with concurrency cap
  const CONCURRENCY = 20;
  for (let i = 0; i < entries.length; i += CONCURRENCY) {
    const batch = entries.slice(i, i + CONCURRENCY);
    const statuses = await Promise.all(
      batch.map((sessionId) => {
        const dir = join(stateDir, sessionId);
        return classifySession(dir, thresholdMs, workspaceRecovery);
      }),
    );
    for (let j = 0; j < batch.length; j++) {
      result[batch[j]] = statuses[j];
    }
  }

  return result;
}

// ---------------------------------------------------------------------------
// Classification — ports classifySession / classifyLiveSession from Go
// ---------------------------------------------------------------------------

async function classifySession(
  dir: string,
  thresholdMs: number,
  workspaceRecovery: boolean,
): Promise<AttentionStatus> {
  const pidResult = await findSessionPID(dir);

  if (pidResult.pid > 0) {
    return classifyLiveSession(dir, thresholdMs);
  }

  // No live process — check last event
  const eventsPath = join(dir, 'events.jsonl');
  const evt = await readLastEvent(eventsPath);
  if (!evt) return 'idle';

  const eventTime = parseEventTime(evt.timestamp);
  if (!eventTime) return 'idle';

  // Stale lock (dead PID) with workspace recovery enabled
  if (pidResult.hasStale && workspaceRecovery) {
    if (Date.now() - eventTime.getTime() > INTERRUPTED_MAX_AGE_MS) {
      return 'idle';
    }
    if (evt.type.startsWith(EVENT_TURN_END) || evt.type.startsWith(EVENT_MESSAGE)) {
      return 'waiting';
    }
    return 'interrupted';
  }

  // No stale lock — original dead-session logic
  if (Date.now() - eventTime.getTime() > DEAD_SESSION_MAX_AGE_MS) {
    return 'idle';
  }
  if (evt.type.startsWith(EVENT_TURN_END) || evt.type.startsWith(EVENT_MESSAGE)) {
    return 'waiting';
  }
  return 'idle';
}

async function classifyLiveSession(
  dir: string,
  thresholdMs: number,
): Promise<AttentionStatus> {
  const eventsPath = join(dir, 'events.jsonl');
  const evt = await readLastEvent(eventsPath);
  if (!evt) return 'stale';

  const eventTime = parseEventTime(evt.timestamp);
  if (eventTime && Date.now() - eventTime.getTime() > thresholdMs) {
    return 'stale';
  }

  // Classify by event type
  if (evt.type.startsWith(EVENT_TURN_END) || evt.type.startsWith(EVENT_MESSAGE)) {
    return 'waiting';
  }
  if (evt.type.startsWith(EVENT_TOOL_EXECUTION)) {
    return 'working';
  }
  if (evt.type.startsWith(EVENT_TURN_START)) {
    return 'thinking';
  }
  if (evt.type === EVENT_COMPACTION) {
    return 'compacting';
  }
  if (
    evt.type.startsWith(EVENT_HOOK) ||
    evt.type.startsWith(EVENT_SUBAGENT) ||
    evt.type.startsWith(EVENT_PLAN_CHANGED) ||
    evt.type.startsWith(EVENT_SKILL_INVOKED)
  ) {
    return 'active';
  }
  if (evt.type === EVENT_SHUTDOWN || evt.type === EVENT_ABORT) {
    return 'idle';
  }
  if (evt.type === EVENT_MODEL_CHANGE || evt.type === EVENT_SYSTEM_MESSAGE) {
    return 'waiting';
  }

  // user.message, session.start, or unknown — AI hasn't started responding
  return 'active';
}

// ---------------------------------------------------------------------------
// PID / Lock file handling
// ---------------------------------------------------------------------------

async function findSessionPID(dir: string): Promise<PidResult> {
  let matches: string[];
  try {
    const entries = await readdir(dir);
    matches = entries
      .filter((name) => name.startsWith('inuse.') && name.endsWith('.lock'))
      .slice(0, MAX_LOCK_FILES_PER_SESSION)
      .map((name) => join(dir, name));
  } catch {
    return { pid: 0, hasStale: false };
  }

  if (matches.length === 0) {
    return { pid: 0, hasStale: false };
  }

  let hasStale = false;

  for (const lockPath of matches) {
    try {
      const info = await stat(lockPath);
      if (!info.isFile()) continue;
      if (info.size >= MAX_LOCK_FILE_SIZE) continue;

      const raw = await readFile(lockPath, 'utf-8');
      const pidStr = raw.trim();

      if (pidStr.length === 0) continue;
      if (!/^\d+$/.test(pidStr)) continue;

      const pid = parseInt(pidStr, 10);
      if (pid <= 0 || pid > MAX_PID) continue;

      if (await isProcessAlive(pid)) {
        return { pid, hasStale };
      }
      hasStale = true;
    } catch {
      continue;
    }
  }

  return { pid: 0, hasStale };
}

/**
 * Check if a process with the given PID is alive.
 * On Windows: uses tasklist to check for the PID.
 * On other platforms: sends signal 0 via process.kill.
 */
async function isProcessAlive(pid: number): Promise<boolean> {
  if (process.platform === 'win32') {
    try {
      // Use tasklist filtered by PID — exits with code 0 and produces output if found
      const { stdout } = await execAsync(
        `tasklist /FI "PID eq ${pid}" /NH /FO CSV`,
        { timeout: 2000 },
      );
      // tasklist outputs "INFO: No tasks are running..." when PID not found
      return stdout.includes(`"${pid}"`);
    } catch {
      return false;
    }
  }

  // Unix: signal 0 checks existence without killing
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// Event file reading — O(1) seek-from-end strategy
// ---------------------------------------------------------------------------

async function readLastEvent(path: string): Promise<SessionEvent | null> {
  let fh;
  try {
    fh = await open(path, 'r');
    const info = await fh.stat();

    if (info.size === 0) return null;

    const chunkSize = Math.min(LAST_CHUNK_SIZE, info.size);
    const offset = info.size - chunkSize;

    const buffer = Buffer.alloc(chunkSize);
    const { bytesRead } = await fh.read(buffer, 0, chunkSize, offset);
    const content = buffer.subarray(0, bytesRead).toString('utf-8');

    // Find the last complete line (skip trailing newlines)
    const trimmed = content.replace(/[\r\n]+$/, '');
    const lastNL = trimmed.lastIndexOf('\n');
    const lastLine = lastNL >= 0 ? trimmed.substring(lastNL + 1) : trimmed;

    if (!lastLine) return null;

    const evt: SessionEvent = JSON.parse(lastLine);
    return evt;
  } catch {
    return null;
  } finally {
    await fh?.close();
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function parseEventTime(ts: string): Date | null {
  if (!ts) return null;
  const date = new Date(ts);
  if (isNaN(date.getTime())) return null;
  return date;
}

function getSessionStatePath(): string {
  const override = process.env.DISPATCH_SESSION_STATE;
  if (override) {
    // Reject UNC paths on Windows
    if (process.platform === 'win32' && override.startsWith('\\\\')) {
      return '';
    }
    return override;
  }
  return join(homedir(), SESSION_STATE_REL);
}
